### ISSUE TITLE
Build instructions updates / changes to build environment

### ISSUE THREAD

**Issue by oam7575** [rescuezilla/rescuezilla#161]
**TL/DR**
* The build instructions appear to need updating
* Possible problems with partclone-nbd ( or i have missed a step )
* Possibly new packages added to assist with building on newer Ubuntu releases
* Change mksquashfs code to use zstd compression resulting in small ISO size
* ISO's appear to be functional after changes - limited testing carried out.
* Change build scripts to use local apt-cacher-ng install for faster package downloads
Apt-cacher-ng changes are _not_ required; however helpful when setting things up from scratch and having a slow connection.

**Begins** 

I am attempting to follow the build guide here : https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md

It appears the build instructions are broken, or out of date when compared to the current code base.

I wish to make the amd64 builds

The work flow indicates _sudo make amd64 i386_
This fails with no target for amd64

Inspecting the Makefile shows targets of i386, focal and groovy

Executing : make i386 ; results in an i386 iso being generated

Executing make focal or make groovy both results in error messages when attempting to compile partclone-nbd

The error message relates to a missing include for stropts.h 
This stackoverflow page : https://stackoverflow.com/questions/61029226/how-do-i-install-libraries-for-stropts-h
indicates that Stropts.h is a STREAMS library and has been deprecated / removed for some time.

removing partclone-nbd from the required applications results in a iso that appears to be functional ( although this has not been thoroughly tested ) 

Assuming the removal of partclone-nbd removes the explore backup function ( acceptable for me for testing )

**Notes on environment**
**OS**
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"

**debootstrap**
debootstrap 1.0.118ubuntu1.3

**Changes**
**debootstrap**
Based on this issue https://github.com/rescuezilla/rescuezilla/issues/72 some research suggested that lxc-libvert or lxc alone might solve the issue.

I installed lxc via aptitude install lxc

This appears to resolve the issue with debootstrap

**Build.sh**
Change build.sh script to use zstd compression resulting in small ISO size

Find this line
mksquashfs chroot image/casper/filesystem.squashfs -e boot -e /sys

Change it to read like so
mksquashfs chroot image/casper/filesystem.squashfs -comp zstd -b 1M -Xcompression-level 10 -e boot -e /sys

Resulting Groovy iso ~785Mb vs ~850Mb for official release ( note missing partclone-nbd will be part of the size reduction also )

Higher compression levels are possible; however the suggested setting should result in an ISO that is small and built faster than with gzip.

**apt-cacher-ng**
These changes are not required but were useful for myself.
Essentially change all instances of http://archive.ubuntu.com/ubuntu to point to my local apt-cacher-ng install
http://local.ip.of.server:3142/ubuntu

**Comment by shasheene** [comment]
Wow thanks @oam7575!

I presume you are [Sourceforge user OAM775](https://sourceforge.net/u/oam775/activity), the [developer of several _Redo Backup and Recovery_ updates between 2014-2016](https://github.com/rescuezilla/rescuezilla/wiki/Bugs-in-unofficial-Redo-Backup-updates#identifying-redo-backup-versions). Thanks for your great work over the years, as I've said [before](https://sourceforge.net/p/redobackup/discussion/general/thread/116063b485/#fe6d/98c4), your releases on the forums helped inspire me to start developing Rescuezilla!

I'm not surprised the "Build without docker" instructions have developed issues. On my local machine I stick to the [Build ISO with docker](https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md#build-iso-with-docker) and use the commands listed in the [.travis.yml](https://github.com/rescuezilla/rescuezilla/blob/master/.travis.yml) file.

By using the `.travis.yml` and `Dockerfile`, the 64-bit Focal and 64-bit Groovy releases build fine (including the `partclone-nbd` executable). The i386 build is currently disabled ([task #150](https://github.com/rescuezilla/rescuezilla/issues/150)). The Dockerfile contains the most up-to-date version of the instructions, and it's what I'm constantly using. The `.travis.yml` instructions are the instructions that get executed on the Travis-CI.org service to produce an ISO image that gets uploaded to GitHub when a release is tagged. Incidentally the Travis-CI.org service is going offline on December 31 2020, so I intend on [switching to a new CI service (task #149)](https://github.com/rescuezilla/rescuezilla/issues/149). For completeness, I note here I am intending on [overhauling the build environment's ability to compile software (task #150)](https://github.com/rescuezilla/rescuezilla/issues/150).

I will take a look at fixing the "Build without docker" instructions.

Thanks for the recommendation of apt-cache-ng. Caching packages between builds is very important. I'l look into LXC, but the most lightweight way to achieve #72 maybe to modify debootstrap to more officially support caching.

Switching to zstd compression of the SquashFS root filesystem is indeed a good idea. I am still hoping to implement a [persistent partition between boots (task #8)](https://github.com/rescuezilla/rescuezilla/issues/8), but as mentioned in the aformentioned task, the most likely approach is to achieve this is to [simply use Rufus to create the USB stick](https://github.com/rescuezilla/rescuezilla/issues/8#issuecomment-717112242). This will mean Rescuezilla would maintain the SquashFS root filesystem approach for the long term.

When modifying the compression level, there is the potential downside of much slower decompression. I really like zstandard compression because it's as fast as GZIP for both compression and decompression, and provides better compression ratios. I wonder if the random access of compressed data issue described in [task #153](https://github.com/rescuezilla/rescuezilla/issues/153) matters here. I think it decompresses the entire squashfs root filesytem to a ramdisk, so probably not.

By the way, I can see from your profile that you've had some experience with Git and GitHub. As you know, you're free to fork the Rescuezilla GitHub repository, add it as a remote., push the changes you made so far, and make a pull request. If you're unfamiliar with this aspect of git, I recommend [this guide](https://rogerdudler.github.io/git-guide/) which may help you add the oam7575-fork remote, then push your changes there. By the way, I really enjoy using the command-line tool called "tig" (that's available in the standard Linux repositories) to view git history. You may too.

**Even if you don't want to create a pull request, simply adding the line `https://github.com/rescuezilla/rescuezilla/issues/161` to your commit message then pushing it to your fork will make this GitHub issue automatically link to it. Doing this will prevent me having to duplicate exactly the work you've done so far: instead I could just fetch your branch and build exactly the codebase that you're working from!**

Thanks again for your feedback here, and on your work on Redo Backup between 2014-2016. Hopefully we can continue to collaborate on ideas and code going forward!

**Comment by oam7575** [comment]
Hi Shasheene,

You are correct that was me on the old RedoBackup threads.

For docker build instructions; I have used docker from time to time and I know its the way forward for a lot of things; but it does my head in most of the time so was actively trying to avoid it.

I'll try and make some time to play with the Docker side of things and see how I go.

The zstd squashfs testing was mostly to see how small the ISO could be compressed to.
There is a saving to be had, but not massive amounts due to the type of data that is being worked with.
Possibly something to be kept in mind, but not at the top of the list.

Potentially using zstd as the default compression engine with options to select pigz for older images might allow better image size vs speed trade off; however it is very much a case by case situation.
Depending on data set zstd compression tends to either be faster at the same size, or much better size at the same speed.
Decompression tends to be faster overall vs most other options.

https://community.centminmod.com/threads/round-4-compression-comparison-benchmarks-zstd-vs-brotli-vs-pigz-vs-bzip2-vs-xz-etc.18669/

apt-cacher-ng is also very much a case by case situation and its something external to the project that needs to be configured before hand.
For an active dev like yourself I would strongly encourage it; for someone at a large organization with highspeed internet and building a single image once, probably not much to worry about. 

As for me making major changes or forking code, that would probably be unlikely at this time.
As I previously indicated I am not using the code myself any more and other things in real life tend to take my time.

With that said; its pleasing to hear nice feed back and its certainly good to see the effort you are going to moving the project forward.

**Comment by shasheene** [comment]
That's fair enough!

By the way, you don't need to _use_ to install and use docker at all. It's just that the most up-to-date build dependencies are listed in the Dockerfile, and the most up-to-date instructions around running `make groovy` and `make focal` (and _not_ running `make i386`) is available in the .travis.yml file. And because these instructions are exercised every single release they are effectively guaranteed to be up-to-date and work -- unlike the standard instructions.

I don't actually believe Docker is the future, and I don't like that it adds another layer of complexity and makes it less likely for other developers to contribute to Rescuezilla. My use of Docker in the Rescuezilla build is just a crutch: It makes it easier to do development across a number of different Linux distributions because the Dockerfile ensures the host build environment is always based upon the same package environment (Ubuntu 18.04 Bionic environment). Running on different package environments may break things due to eg, the problems described by [task #150](https://github.com/rescuezilla/rescuezilla/issues/150).

The proper solution is to make the build robustly work across many more package environments, which is something I am moving towards (but it will take completing all the GitHub tasks I have linked to in this discussion to achieve).

**Comment by shasheene** [comment]
One more thing I should mention. Many advanced users like yourself may be interested in [task #119 ("Rescuezilla as a Standalone Ubuntu Package")](https://github.com/rescuezilla/rescuezilla/issues/119). There actually is a 250 kilobyte deb file in the Assets section at the bottom of the [releases page](http://github.com/rescuezilla/rescuezilla/latest).

The standalone deb can be easily installed in any Debian/Ubuntu environment. (though I haven't yet configured the virtualenv so that the Python dependencies are automatically installed). If you (or any other developer reading this) just wants to hack at the Python application itself and are comfortable with figuring out how to use pip3 to install the dependencies, then the standalone deb file may be the way to go.

**Comment by mzky** [comment]
> **TL/DR**
> 
> * 构建说明似乎需要更新
> * partclone-nbd 可能存在的问题（或者我错过了一个步骤）
> * 可能添加了新软件包以帮助构建较新的 Ubuntu 版本
> * 更改 mksquashfs 代码以使用 zstd 压缩导致 ISO 大小变小
> * ISO 在更改后似乎可以正常工作 - 进行了有限的测试。
> * 变化构建脚本使用本地的apt-cacher的-NG安装更快的下载包
>   的apt-cacher的-ng的变化是_不是_必需的; 但是在从头开始设置并且连接速度较慢时很有帮助。
> 
> **开始**
> 
> 我正在尝试遵循此处的构建指南：[https](https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md) : [//github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md](https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md)
> 
> 与当前代码库相比，构建指令似乎已损坏或已过时。
> 
> 我希望进行 amd64 构建
> 
> 工作流程表明_sudo make amd64 i386_
> 这失败了，没有 amd64 的目标
> 
> 检查 Makefile 显示 i386、focal 和 groovy 的目标
> 
> 执行 : make i386 ; 导致生成 i386 iso
> 
> 尝试编译 partclone-nbd 时，执行 make focus 或 make groovy 都会导致错误消息
> 
> 该错误消息与 stropts.h 缺少包含有关
> 此 stackoverflow 页面：[https](https://stackoverflow.com/questions/61029226/how-do-i-install-libraries-for-stropts-h)
> ://stackoverflow.com/questions/61029226/how-do-i-install-libraries-for-stropts-h 表示 Stropts.h 是STREAMS 库，并且已经弃用/删除了一段时间。
> 
> 从所需的应用程序中删除 partclone-nbd 会产生一个看起来可以正常工作的 iso（尽管这还没有经过彻底测试）
> 
> 假设删除 partclone-nbd 删除了 explore 备份功能（我可以接受测试）
> 
> ****
> **OS****环境注意事项**
> DISTRIB_ID=Ubuntu
> DISTRIB_RELEASE=20.04
> DISTRIB_CODENAME=focal
> DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
> 
> **debootstrap**
> debootstrap 1.0.118ubuntu1.3
> 
> **更改**
> **debootstrap**
> 基于此问题[#72，](https://github.com/rescuezilla/rescuezilla/issues/72)一些研究表明单独使用 lxc-libvert 或 lxc 可能会解决该问题。
> 
> 我通过 aptitude install lxc 安装了 lxc
> 
> 这似乎解决了 debootstrap 的问题
> 
> **Build.sh**
> 更改 build.sh 脚本以使用 zstd 压缩导致较小的 ISO 大小
> 
> 找到这一行
> mksquashfs chroot image/casper/filesystem.squashfs -e boot -e /sys
> 
> 将其更改为
> mksquashfs chroot image/casper/filesystem.squashfs -comp zstd -b 1M -Xcompression-level 10 -e boot -e /sys
> 
> 结果 Groovy iso ~785Mb vs ~850Mb 正式发布（注意缺少 partclone-nbd 也将是尺寸减小的一部分）
> 
> 更高的压缩级别是可能的；然而，建议的设置应该会导致 ISO 比 gzip 更小且构建速度更快。
> 
> **apt-cacher-ng**
> 这些更改不是必需的，但对我自己很有用。
> 基本上将[http://archive.ubuntu.com/ubuntu 的](http://archive.ubuntu.com/ubuntu)所有实例更改为指向我本地的 apt-cacher-ng install
> http://local.ip.of.server:3142/ubuntu




> **TL/DR**
> 
> * The build instructions appear to need updating
> * Possible problems with partclone-nbd ( or i have missed a step )
> * Possibly new packages added to assist with building on newer Ubuntu releases
> * Change mksquashfs code to use zstd compression resulting in small ISO size
> * ISO's appear to be functional after changes - limited testing carried out.
> * Change build scripts to use local apt-cacher-ng install for faster package downloads
>   Apt-cacher-ng changes are _not_ required; however helpful when setting things up from scratch and having a slow connection.
> 
> **Begins**
> 
> I am attempting to follow the build guide here : https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md
> 
> It appears the build instructions are broken, or out of date when compared to the current code base.
> 
> I wish to make the amd64 builds
> 
> The work flow indicates _sudo make amd64 i386_
> This fails with no target for amd64
> 
> Inspecting the Makefile shows targets of i386, focal and groovy
> 
> Executing : make i386 ; results in an i386 iso being generated
> 
> Executing make focal or make groovy both results in error messages when attempting to compile partclone-nbd
> 
> The error message relates to a missing include for stropts.h
> This stackoverflow page : https://stackoverflow.com/questions/61029226/how-do-i-install-libraries-for-stropts-h
> indicates that Stropts.h is a STREAMS library and has been deprecated / removed for some time.
> 
> removing partclone-nbd from the required applications results in a iso that appears to be functional ( although this has not been thoroughly tested )
> 
> Assuming the removal of partclone-nbd removes the explore backup function ( acceptable for me for testing )
> 
> **Notes on environment**
> **OS**
> DISTRIB_ID=Ubuntu
> DISTRIB_RELEASE=20.04
> DISTRIB_CODENAME=focal
> DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
> 
> **debootstrap**
> debootstrap 1.0.118ubuntu1.3
> 
> **Changes**
> **debootstrap**
> Based on this issue #72 some research suggested that lxc-libvert or lxc alone might solve the issue.
> 
> I installed lxc via aptitude install lxc
> 
> This appears to resolve the issue with debootstrap
> 
> **Build.sh**
> Change build.sh script to use zstd compression resulting in small ISO size
> 
> Find this line
> mksquashfs chroot image/casper/filesystem.squashfs -e boot -e /sys
> 
> Change it to read like so
> mksquashfs chroot image/casper/filesystem.squashfs -comp zstd -b 1M -Xcompression-level 10 -e boot -e /sys
> 
> Resulting Groovy iso ~785Mb vs ~850Mb for official release ( note missing partclone-nbd will be part of the size reduction also )
> 
> Higher compression levels are possible; however the suggested setting should result in an ISO that is small and built faster than with gzip.
> 
> **apt-cacher-ng**
> These changes are not required but were useful for myself.
> Essentially change all instances of http://archive.ubuntu.com/ubuntu to point to my local apt-cacher-ng install
> http://local.ip.of.server:3142/ubuntu

hi oam7575!
I tried compiling under Debian 10.3 ARM64. Error message: " Unmet build dependencies: devscripts"
I tried installing dev-scripts using NPM and executed "aptitude install lxc". The same problem still arises.
Have you ever encountered such a situation?
I don't know if the ARM64 platform is the reason.

![image](https://user-images.githubusercontent.com/13345233/120947867-61bee900-c773-11eb-9f7e-87175d8d9513.png)

cpu:
root@super-PC:~# cat /proc/cpuinfo
processor       : 0
model name      : HUAWEI Kunpeng 920 2249K
BogoMIPS        : 200.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt lpae evtstrm
CPU implementer : 0x48
CPU architecture: 8
CPU variant     : 0x1
CPU part        : 0xd01
CPU revision    : 0

**Comment by mzky** [comment]
I just found an update two days ago，but the problem is the same.
![image](https://user-images.githubusercontent.com/13345233/120949318-d182a300-c776-11eb-94e8-81cf256cdc32.png)

**Comment by shasheene** [comment]
Hi @mzky  Can you please copy your previous posts and create a new GitHub issue (then delete the posts above). This GitHub issue is specifically for OAM7575's build _enhancements_ and recommendations like switching to zstandard compression.

**Comment by oam7575** [comment]
Based on the July 2021 status of the build instructions and my work for pull request #245 [](https://github.com/rescuezilla/rescuezilla/pull/245) I would be happy to call this one closed.

With pull #245 I have made a change to the build instructions regarding libntfs3g88.
It appears there is either a type-o or version change and libntfs-3g88 needs to be modified to **libntfs-3g883**

With that fix and the other minor adjustments I would call this one done for now.

**Comment by shasheene** [comment]
Sounds good. Though the libntfs-3g883 version number change is not yet required on the package environment created by the Dockerfile. The package name change will eventually be adopted once the host build environment is updated.
